src/transform.rs

```diff
@@ -48,9 +48,100 @@ where
     }
     html.push_str(&escape_html(&text[cursor..])?);
 
+    println!("{}", html);
+
     Ok(Builder::default()
         .tags(hashset!["a", "p", "br"])
         .link_rel(Some("noopener nofollow"))
         .clean(&html)
         .to_string())
 }
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    fn passes_through_text() {
+        assert_eq!(bio("foo", |_, _| Ok(None)).unwrap(), "foo");
+        assert_eq!(bio("foo:bar", |_, _| Ok(None)).unwrap(), "foo:bar");
+    }
+
+    #[test]
+    fn escapes_html_characters() {
+        assert_eq!(bio("<>&", |_, _| Ok(None)).unwrap(), "&lt;&gt;&amp;");
+        assert_eq!(
+            bio("<a></a>", |_, _| Ok(None)).unwrap(),
+            "&lt;a&gt;&lt;/a&gt;"
+        );
+    }
+
+    #[test]
+    fn converts_newlines_to_br() {
+        assert_eq!(bio("\n", |_, _| Ok(None)).unwrap(), "<br>");
+        assert_eq!(bio("\r\n", |_, _| Ok(None)).unwrap(), "<br>");
+    }
+
+    #[test]
+    fn converts_links_to_a_tags() {
+        assert_eq!(
+            bio("https://example.com", |_, _| Ok(None)).unwrap(),
+            "<a href=\"https://example.com\" rel=\"noopener nofollow\">https://example.com</a>"
+        );
+        assert_eq!(
+            bio("http://example.com", |_, _| Ok(None)).unwrap(),
+            "<a href=\"http://example.com\" rel=\"noopener nofollow\">http://example.com</a>"
+        );
+        assert_eq!(
+            bio("http://‽.com/∰/", |_, _| Ok(None)).unwrap(),
+            "<a href=\"http://‽.com/∰/\" rel=\"noopener nofollow\">http://‽.com/∰/</a>"
+        );
+    }
+
+    #[test]
+    fn converts_hashtags_to_links() {
+        // TODO: we don't have hashtags atm, so we just fake-link them!
+        assert_eq!(
+            bio("#hashtag", |_, _| Ok(None)).unwrap(),
+            "<a href=\"#\" rel=\"noopener nofollow\">#hashtag</a>"
+        );
+    }
+
+    #[test]
+    fn converts_mentions_to_links() {
+        use std::env;
+        env::set_var("DOMAIN", "localhost"); // TODO: this is bad and should go away, _somehow_.
```

Comment:
I _really_ don't like this. There's a very annoying line between "making things easily globally configurable" and _testing_ those globally-configurable things, and I'm worried using `lazy_static!` and env vars was a Bad Choice, or Something v_v

Reply:
yeaaaaah this is gross but it's gonna be the way it is unless we rejigger a Lot of stuff
so
yeah
let's just... ignore it, mmmkay?